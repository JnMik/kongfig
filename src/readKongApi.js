import kongState from './kongState';

export default async (adminApi) => {
    return Promise.all([kongState(adminApi), adminApi.fetchPluginSchemas()])
        .then(([state, schemas]) => {
            const prepareConfig = (plugin, config) => stripConfig(config, schemas.get(plugin));
            const parseApiPluginsForSchemes = plugins => parseApiPlugins(plugins, prepareConfig);

            return {
                apis: parseApis(state.apis, parseApiPluginsForSchemes),
                consumers: parseConsumers(state.consumers)
            }
        })
};

function parseConsumers(consumers) {
    return consumers.map(({username, credentials, acls, ..._info}) => {
        return {
            username,
            _info,
            acls: acls.map(({group, ..._info}) => ({group, _info})),
            credentials: zip(Object.keys(credentials), Object.values(credentials))
                .map(parseCredential)
                .reduce((acc, x) => acc.concat(x), [])
        };
    });
}

function zip(a, b) {
    return a.map((n, index) => [n, b[index]]);
}

function parseCredential([credentialName, credentials]) {
    return credentials.map(({consumer_id, id, created_at, ...attributes}) => {
        return {
            name: credentialName,
            attributes,
            _info: {id, consumer_id, created_at}
        }
    });
}

function parseApis(apis, parseApiPlugins) {
    return apis.map(({
        name, plugins,
        request_host, request_path, strip_request_path, preserve_host, upstream_url,
        id, created_at}) => {
        return {
            name,
            plugins: parseApiPlugins(plugins),
            attributes: {
                request_host,
                request_path,
                strip_request_path,
                preserve_host,
                upstream_url,
            },
            _info: {
                id,
                created_at
            }
        };
    });
}

function parseApiPlugins(plugins, prepareConfig) {
    return plugins.map(({
        name,
        config,
        id, api_id, consumer_id, enabled, created_at
    }) => {
        return {
            name,
            attributes: {
                config: prepareConfig(name, config)
            },
            _info: {
                id,
                //api_id,
                consumer_id,
                enabled,
                created_at
            }
        };
    });
}

function stripConfig(config, schema) {
    const mutableConfig = {...config};

    if (false) {
        // strip default keys
        Object.keys(config).map(key => {
            if (schema[key].hasOwnProperty('default') && schema[key].default === config[key]) {
                delete mutableConfig[key];
            }
        });
    }

    // remove some cache values
    delete mutableConfig['_key_der_cache'];
    delete mutableConfig['_cert_der_cache'];

    return mutableConfig;
}
