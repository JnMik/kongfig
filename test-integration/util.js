import adminApi from '../lib/adminApi';
import readKongApi from '../lib/readKongApi';
import execute from '../lib/core';
import invariant from 'invariant';
import pad from 'pad';

invariant(process.env.TEST_INTEGRATION_KONG_HOST, `
    Please set ${'TEST_INTEGRATION_KONG_HOST'.bold} env variable

    TEST_INTEGRATION_KONG_HOST=localhost:8001 yarn test

    ${'WARNING! Running integration tests are going to remove all data from the kong'.red.bold}.
`);

const UUIDRegex = /[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[89ab][a-f0-9]{3}-?[a-f0-9]{12}/;
let uuids = {};
let log = [];

export const testAdminApi = adminApi({
    host: process.env.TEST_INTEGRATION_KONG_HOST,
    https: false,
    ignoreConsumers: false,
    cache: false,
});

export const getLog = () => log;
export const logger = message => {
    const m = cloneObject(message);

    if (m.hasOwnProperty('uri')) {
        m.uri = m.uri.replace(process.env.TEST_INTEGRATION_KONG_HOST, 'localhost:8001');
    }

    log.push(ignoreKeys(m, ['created_at']));
};

const _ignoreKeys = (obj, keys) => {
    if (obj instanceof Array) {
        obj.forEach((item) => _ignoreKeys(item, keys));
    } else if (typeof obj === 'object') {
        Object.getOwnPropertyNames(obj).forEach(key => {
            if (typeof obj[key] === 'string' && obj[key].match(UUIDRegex)) {
                const uuid = obj[key].match(UUIDRegex)[0];

                if (!uuids.hasOwnProperty(obj[uuid])) {
                    const id = pad(12, `${Object.keys(uuids).length + 1}`, '0');
                    uuids[uuid] = `2b47ba9b-761a-492d-9a0c-${id}`;
                }

                obj[key] = obj[key].replace(uuid, uuids[uuid]);
            } else if (keys.indexOf(key) !== -1) {
                obj[key] = `___${key}___`;
            } else {
                _ignoreKeys(obj[key], keys);
            }
        });
    }

    return obj;
};

const cloneObject = obj => JSON.parse(JSON.stringify(obj));

export const ignoreKeys = (message, keys) => _ignoreKeys(cloneObject(message), keys);

const cleanupKong = async () => {
    const results = await readKongApi(testAdminApi);
    await execute({
        apis: results.apis.map(api => ({ ...api, ensure: 'removed' })),
        consumers: results.consumers.map(consumer => ({ ...consumer, ensure: 'removed' })),
        plugins: results.plugins.map(plugin => ({ ...plugin, ensure: 'removed' })),
    }, testAdminApi);
};

export const tearDown = async () => {
    uuids = {};
    log = [];
    await cleanupKong();
};
