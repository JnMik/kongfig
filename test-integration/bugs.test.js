import execute from '../lib/core';
import { testAdminApi, logger, ignoreKeys, getLog, tearDown } from './util';
import readKongApi from '../lib/readKongApi';

beforeEach(tearDown);

it('should allow updating a global plugin with no attributes', async () => {
    const config = {
        plugins: [{
            name: "cors",
            attributes: {
                enabled: true
            }
        }]
    };

    await execute(config, testAdminApi, logger);
    await execute(config, testAdminApi, logger);
    const kongState = await readKongApi(testAdminApi);

    expect(getLog()).toMatchSnapshot();
    expect(ignoreKeys(kongState, ['created_at'])).toMatchSnapshot();
});
