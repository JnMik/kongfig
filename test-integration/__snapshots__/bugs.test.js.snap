// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow updating a global plugin with no attributes 1`] = `
Array [
  Object {
    "params": Object {
      "body": Object {
        "enabled": true,
        "name": "cors",
      },
      "endpoint": Object {
        "name": "plugins",
        "params": Object {
          "pluginName": "cors",
        },
      },
      "method": "POST",
      "type": "add-global-plugin",
    },
    "type": "action",
  },
  Object {
    "params": Object {
      "body": Object {
        "enabled": true,
        "name": "cors",
      },
      "endpoint": Object {
        "name": "plugins",
        "params": Object {
          "pluginName": "cors",
        },
      },
      "method": "POST",
      "type": "add-global-plugin",
    },
    "type": "request",
    "uri": "http://localhost:8001/plugins",
  },
  Object {
    "ok": true,
    "params": Object {
      "body": Object {
        "enabled": true,
        "name": "cors",
      },
      "endpoint": Object {
        "name": "plugins",
        "params": Object {
          "pluginName": "cors",
        },
      },
      "method": "POST",
      "type": "add-global-plugin",
    },
    "status": 201,
    "statusText": "Created",
    "type": "response",
    "uri": "http://localhost:8001/plugins",
  },
  Object {
    "content": Object {
      "config": Object {
        "credentials": false,
        "preflight_continue": false,
      },
      "created_at": "___created_at___",
      "enabled": true,
      "id": "2b47ba9b-761a-492d-9a0c-000000000001",
      "name": "cors",
    },
    "type": "response-content",
  },
  Object {
    "params": Object {
      "noop": true,
      "plugin": Object {
        "attributes": Object {
          "enabled": true,
        },
        "name": "cors",
      },
      "type": "noop-global-plugin",
    },
    "type": "action",
  },
]
`;

exports[`should allow updating a global plugin with no attributes 2`] = `
Object {
  "apis": Array [],
  "consumers": Array [],
  "plugins": Array [
    Object {
      "_info": Object {
        "created_at": "___created_at___",
        "id": "2b47ba9b-761a-492d-9a0c-000000000002",
      },
      "attributes": Object {
        "config": Object {
          "credentials": false,
          "preflight_continue": false,
        },
        "enabled": true,
      },
      "name": "cors",
    },
  ],
}
`;
